soccer/src/soccer/control/motion_control.cpp

```diff
@@ -0,0 +1,220 @@
+DEFINE_INT64(params::kMotionControlParamModule, rotation_windup, 0,
+             "Windup limit for rotation (unknown units)");
+DEFINE_FLOAT64(params::kMotionControlParamModule, translation_kp, 2.0,
+               "Kp for translation ((m/s)/m)");
+DEFINE_FLOAT64(params::kMotionControlParamModule, translation_ki, 0.0,
+               "Ki for translation ((m/s)/(m*s))");
+DEFINE_FLOAT64(params::kMotionControlParamModule, translation_kd, 0.0,
+               "Kd for translation ((m/s)/(m/s))");
+DEFINE_INT64(params::kMotionControlParamModule, translation_windup, 0,
+             "Windup limit for translation (unknown units)");
+
+MotionControl::MotionControl(int shell_id, rclcpp::Node* node)
+    : shell_id_(shell_id),
+      angle_controller_(0, 0, 0, 50, 0),
+      drawer_(
+          node->create_publisher<rj_drawing_msgs::msg::DebugDraw>(viz::topics::kDebugDrawPub, 10),
+          fmt::format("motion_control/{}", std::to_string(shell_id))) {
+    motion_setpoint_pub_ = node->create_publisher<MotionSetpoint::Msg>(
+        topics::motion_setpoint_pub(shell_id_), rclcpp::QoS(1));
+    // Update motion control triggered on world state publish.
+    trajectory_sub_ = node->create_subscription<planning::Trajectory::Msg>(
+        planning::topics::trajectory_pub(shell_id), rclcpp::QoS(1),
+        [this](planning::Trajectory::Msg::SharedPtr trajectory) {  // NOLINT
+            trajectory_ = rj_convert::convert_from_ros(*trajectory);
+        });
+    world_state_sub_ = node->create_subscription<WorldState::Msg>(
+        vision_filter::topics::kWorldStatePub, rclcpp::QoS(1),
+        [this](WorldState::Msg::SharedPtr world_state_msg) {  // NOLINT
+            RobotState state =
+                rj_convert::convert_from_ros(world_state_msg->our_robots.at(shell_id_));
+
+            // TODO(Kyle): Handle the joystick-controlled case here. In the long run we want to
+            // convert this to an action. Should we do that now?
+            bool is_joystick_controlled = false;
+            MotionSetpoint setpoint;
+            run(state, trajectory_, game_state_, is_joystick_controlled, &setpoint);
+            motion_setpoint_pub_->publish(rj_convert::convert_to_ros(setpoint));
+        });
+    game_state_sub_ = node->create_subscription<GameState::Msg>(
+        referee::topics::kGameStatePub, rclcpp::QoS(1).transient_local(),
+        [this](GameState::Msg::SharedPtr game_state_msg) {  // NOLINT
+            game_state_ = rj_convert::convert_from_ros(*game_state_msg).state;
+        });
+}
+
+void MotionControl::run(const RobotState& state, const planning::Trajectory& trajectory,
+                        const GameState::State& game_state, bool is_joystick_controlled,
+                        MotionSetpoint* setpoint) {
+    // If we don't have a setpoint (output velocities) or we're under joystick
+    // control, reset our PID controllers and exit (but don't force a stop).
+    if ((setpoint == nullptr) || is_joystick_controlled) {
```

Comment:
When will we not have a setpoint but want to keep moving (or I guess not execute a stop) aside from joystick control? Edit: O.K., never mind I think this makes sense.